### PR TITLE
[FEM] Allow modal adding/removal of geometric entities to constraints

### PR DESCRIPTION
**Pull request by AjinkyaDahale**
With these changes applied, if nothing is selected while a constraint is being edited, pressing the add/remove button will put the user in an add/remove mode so that on clicking a face/edge in the viewport, that edge/face will be added to/removed from the list of entities to be constrained with the given FEM constraint.

Currently restricted to displacement constraints. I plan to add this to the rest of the constraints and make the behavior consistent with task-views in other workbenches which add/remove boundary entities.

### CODE REVIEW

**Comment by AjinkyaDahale** [comment]
@berndhahnebach did you get the chance to test this?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5117">pipeline status</a> for feature branch [PR_5117](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5117). Pipeline [417926682](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/417926682) was triggered at [4398688](https://github.com/FreeCAD/FreeCAD/pull/5117/commits/43986880469341978e5efdc9488784c09081ed8c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5117">pipeline status</a> for feature branch [PR_5117](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_5117). Pipeline [419732486](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/419732486) was triggered at [5bd83fe](https://github.com/FreeCAD/FreeCAD/pull/5117/commits/5bd83fe86f93e215dbc593bc8e5732f5796086b8). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
> @berndhahnebach did you get the chance to test this?

not yet, I am overwhelmed with BIM related work ... luckily some opensource coding too. But non for FEM ..

Moved it on top of FEM todo ... cheers bernd

File: src/Mod/Fem/Gui/TaskFemConstraint.h [hunk 1]
```diff
@@ -59,12 +59,11 @@ protected Q_SLOTS:
 
 protected:
     virtual void changeEvent(QEvent *e) { TaskBox::changeEvent(e); }
+    const QString makeRefText(const std::string& objName, const std::string& subName) const;
     const QString makeRefText(const App::DocumentObject* obj, const std::string& subName) const;
     virtual void keyPressEvent(QKeyEvent * ke);
     void createDeleteAction(QListWidget* parentList);
     bool KeyEvent(QEvent *e);
-
-private:
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskFemConstraint.h:67]: Why this change?

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
+/***************************************************************************
+ *   Copyright (c) 2021 FreeCAD Developers                                 *
+ *   Authors: Ajinkya Dahale                                               *
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:3]: missing mail address

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+# include <sstream>
+#endif
+
+#include "TaskFemConstraintOnBoundary.h"
+#include <App/Application.h>
+#include <Base/Tools.h>
+#include <Gui/Command.h>
+#include <Gui/Selection.h>
+#include <Gui/SelectionFilter.h>
+// #include <Mod/Part/App/PartFeature.h>
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:52]: why these comments?

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+void TaskFemConstraintOnBoundary::_addToSelection(bool checked)
+{
+    if (checked)
+    {
+        const auto& selection = Gui::Selection().getSelectionEx(); //gets vector of selected objects of active document
+        if (selection.empty()) {
+            this->clearButtons(refAdd);
+            selChangeMode = refAdd;
+            // TODO: set up depending on selChangeMode
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:78]: Do you plan this as separate PR? (a ToDo in new files is somehow unusual ;-) )

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
+/***************************************************************************
+ *   Copyright (c) 2021 FreeCAD Developers                                 *
+ *   Authors: Ajinkya Dahale                                               *
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.h:3]: missing mail address

**Review by donovaly — COMMENTED** [review]
Looks good. I only have questions to understand

Please rebase

File: src/Mod/Fem/Gui/TaskFemConstraint.h (hunk 1 above)
Comment (AjinkyaDahale) [reply src/Mod/Fem/Gui/TaskFemConstraint.h:67]: Made the method protected so that sub-classes can use it.

Comment (donovaly) [reply src/Mod/Fem/Gui/TaskFemConstraint.h:67]: Ah sure, stupid me.

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp (hunk 3 above)
Comment (AjinkyaDahale) [reply src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp]: Remnants of files I thought I needed. Either they were included in from elsewhere or not needed.

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp (hunk 4 above)
Comment (AjinkyaDahale) [reply src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:76]: This particular one was not necessary, I believe. It was just to make sure add and remove did different things when I used the same template for both.

Comment (AjinkyaDahale) [reply src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:76]: In general I leave some TODO's if I believe there are places where improvements can be made but I don't have the time or any other reason.

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+void TaskFemConstraintOnBoundary::onSelectionChanged(const Gui::SelectionChanges&msg)
+{
+    if (msg.Type == Gui::SelectionChanges::AddSelection) {
+        switch (selChangeMode) {
+        case refAdd:
+            // TODO: Optimize to just perform actions on the newly selected item
+            this->addToSelection();
+            // ui->lw_references->addItem(makeRefText(msg.pObjectName, msg.pSubName));
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:109]: this comment can go

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        default:
+            return;
+        }
+    }
+}
+
+void TaskFemConstraintOnBoundary::exitSelectionChangeMode()
+{
+    selChangeMode = none;
+    // Gui::Selection().rmvSelectionGate();
```
Comment (donovaly) [inline src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:125]: this too

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp (hunk 6 above)
Comment (AjinkyaDahale) [reply src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:109]: I don't think so. The TODO is there to suggest that the selection changes can be used to reduce the computational effort, and the commented line suggests how to to do it.

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp (hunk 7 above)
Comment (AjinkyaDahale) [reply src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:125]: yes I guess this one can

File: src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp (hunk 6 above)
Comment (donovaly) [reply src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:109]: OK. However for an outsider, the commented code should follow the Todo to make this clear

Comment (AjinkyaDahale) [reply src/Mod/Fem/Gui/TaskFemConstraintOnBoundary.cpp:108]: OK done!

**Comment by donovaly** [comment]
Many thanks for this feature!

It was a while on my ToDo list to have uniform FEM constraint dialogs that provide the multi-selection feature I once implemented for the Fillet/Chamfer dialog. With this PR this is achievable. 

The only thing I miss is the highlighting of the selected/added elements. Can you please add this in a separate PR?
Another small bit is the title of the widget. Currently it is this:
FreeCAD_lyvwW1eaE9
but it should in this case be e.g. "Displacement constraint parameters".

I will now merge. Can you please note the new feature in the release notes and update the corresponding FEM Wiki page?

**Comment by AjinkyaDahale** [comment]
> Many thanks for this feature!
> 
> It was a while on my ToDo list to have uniform FEM constraint dialogs that provide the multi-selection feature I once implemented for the Fillet/Chamfer dialog. With this PR this is achievable.

Cheers!

> The only thing I miss is the highlighting of the selected/added elements. Can you please add this in a separate PR?

I'll take a look. Do you happen to know somewhere where this already works?

> Another small bit is the title of the widget. Currently it is this: [FEM constraint parameters] but it should in this case be e.g. "Displacement constraint parameters".

Sounds like a straightforward fix once I figure out where to edit.

> I will now merge. Can you please note the new feature in the release notes and update the corresponding FEM Wiki page?

OK.

**Comment by donovaly** [comment]
> > Many thanks for this feature!
> > The only thing I miss is the highlighting of the selected/added elements. Can you please add this in a separate PR?
> 
> I'll take a look. Do you happen to know somewhere where this already works?

Take the PD Fillet or Chamfer dialog.

**Comment by AjinkyaDahale** [comment]
> The only thing I miss is the highlighting of the selected/added elements. Can you please add this in a separate PR?

@donovaly I'm trying to add this following how it's done in PD dress-up in a separate branch: https://github.com/AjinkyaDahale/FreeCAD/tree/fem-constrain-qol-2, but I'm getting a confusing `undefined reference` error. I'll elaborate on it sometime tomorrow but if you can find the error before please let me know.
